grove/connectors/fleetdm/api.py

```diff
@@ -0,0 +1,181 @@
+            # volume - much more than required. We define a jmespath query string to filter
+            # the response from the API down to just the fields we need. The default if none
+            # is set is "*" which returns the whole API response
+            for host in result.body.get("hosts"):
+                filteredResults.append(jmespath.search(jmespath_queries,host))
+                cursor = host.get("software_updated_at")
+
+        # Return the cursor of the last processed date and the results to allow the caller to page as required.
+        return AuditLogEntries(cursor=cursor, entries=filteredResults)
+
+
+
+    def get_software(
+        self,
+        params: Dict[str, Any],                                     # get parameters json dict from https://fleetdm.com/docs/rest-api/rest-api#list-hosts
+        jmespath_queries: str,
+        api_uri: str,
+        cursor: Optional[int],
+        pointer: datetime
+    ) -> AuditLogEntries:
+        jmespath_queries = jmespath_queries
+        
+        """Fetches a list of audit logs which match the provided filters.
+        :return: AuditLogEntries object containing a pagination cursor, and log entries.
+        """
+
+        # Set the page value to the current cursor
+        params["page"] = str(cursor)
+
+        # See psf/requests issue #2651 for why we can happily pass in None values and
+        # not have the request key added to the URI.
+        result = self._get(
+            f"{api_uri}/api/v1/fleet/software/versions",
+            params,
+        )
+
+        filteredResults = []
+        # FleetDM returns an empty software array if there's no more pages of results,
+        # so swap this for a None object
+        # Otherwise, increment the page and continue
+        if int(len((result.body.get("software",[])))) == 0:
+            cursor = None
+        elif cursor is not None:
+            cursor = int(cursor) + 1
+            # The default response for a Fleet Software call including software returns a large
+            # volume - much more than required. We define a jmespath query string to filter
+            # the response from the API down to just the fields we need. The default if none
+            # is set is "*" which returns the whole API response
+            updated_at = result.body.get("counts_updated_at")
+            # There's no way in the REST api to ask for only software that has been updated since
+            # the last time the software table has last been refreshed. So we manually check
+            # the counts_updated_at value against the pointer stored as the last processed time
+            # to see if we should log the software again
+            if(datetime.fromisoformat(str(updated_at)) > datetime.fromisoformat(str(pointer))):
+                for software in result.body.get("software"):
+                    s = jmespath.search(jmespath_queries,software)
```

Comment:
Generally, we avoid post-processing data in the connector and instead use Processors for doing this sort of result filtering.

In the interest of getting this connector shipped we'll merge with this present, but we may need to rebase in future to remove this code in favour of a more generalised processor to handle this :)